hw3/01-characters.js

```diff
@@ -1,4 +1,59 @@
-// url for the Thrones API
-const url = 'https://thronesapi.com/api/v2/Characters';
+/** Sample data from the API
+ * {
+    "id": 47,
+    "firstName": "Wylis",
+    "lastName": "Hodor",
+    "fullName": "Hodor",
+    "title": "Servant of House Stark",
+    "family": "Stark",
+    "image": "hodor.jpg",
+    "imageUrl": "https://thronesapi.com/assets/images/hodor.jpg"
+    },
+ */
 
-fetch(url);
+const url = "https://thronesapi.com/api/v2/Characters";
+
+const resultContainer = document.querySelector("#results");
+
+const fetchData = async (url) => {
+  try {
+    const response = await fetch(url);
+    const characterList = await response.json();
+
+    characterList.forEach(({ firstName, lastName, title, imageUrl }) => {
+      const fullName = `${firstName} ${lastName}`;
+
+      const contentContainer = document.createElement("figcaption");
+
+      const fullNameEle = document.createElement("p");
+      fullNameEle.innerText = fullName;
+      fullNameEle.className = "character-name";
+
+      const titleEle = document.createElement("p");
+      titleEle.innerText = title;
+      titleEle.className = "character-title";
+
+      contentContainer.appendChild(fullNameEle);
+      contentContainer.appendChild(titleEle);
+
+      const img = document.createElement("img");
+      img.src = imageUrl;
+      img.alt = fullName;
+
+      const container = document.createElement("li");
+      container.className = "character-figure";
+
+      container.appendChild(img);
+      container.appendChild(contentContainer);
+
+      resultContainer.appendChild(container);
+      resultContainer.classList.add("grid");
+    });
```

Comment:
> This is a lot of code to put inside a forEach().

This makes sense, but logic is linear and there is only like 25 lines so I don't think it's that much ❓ 

> Consider creating an additional function which would be responsible for creating and appending elements "characters" to the DOM.

Hmmm, how would you do this ❓ I'm seeing the whole chunk is about creating and appending elements "characters", so we'll have to bring everything into another function, which will also make that function too long

If we're splitting each element creation like `li`, `grid`, etc into their own function then it would create too much overhead for just 25 lines I belive

Reply:
That new function would have the dedicated task of taking the data and processing it into into elements to add to the DOM. The function name would match it's responsibility too, for example addCharacterToDom(). So in the future when someone wants to change how the data is displayed on the page it would be easy for them to know to focus on that function instead of looking through the forEach and modifying things in there. It's more about separation of concerns and making it readible. This would mean splitting the individual element creation across different functions, it's just that there would be a function for character creation and presentation.

Reply:
sure!